accessing module variable in separate compilation throws linker error

> a.f90

```fortran
program test
use :: xx
print *, test_xx
end program test
```

> b.f90

```fortran
module xx
integer :: test_xx
end module xx
```

```console
% lfortran b.f90 -c --generate-object-code
% nm b.o
0000000000000000 S _test_xx
0000000000000000 t ltmp0
0000000000000000 s ltmp1
% lfortran a.f90 --generate-object-code
Undefined symbols for architecture arm64:
  "_test_xx", referenced from:
      _main in a.tmp.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
The command 'clang -o a.out /var/folders/r9/27d27cj93ld8zjgr3chdh72h0000gn/T/a.tmp.o  -L"/Users/pranavchiku/repos/lfortran/src/bin/../runtime" -Wl,-rpath,"/Users/pranavchiku/repos/lfortran/src/bin/../runtime" -llfortran_runtime -lm' failed.
Tip: If there is a linker issue, switch the linker using --linker=<CC> option or create an environment variable `export LFORTRAN_LINKER=<CC>`, where CC is clang or gcc
Also, if required use --linker-path=<PATH>, where PATH has location to look for the linker execuatable
```

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -2397,7 +2397,8 @@ RUN(NAME separate_compilation_10 LABELS gfortran llvm EXTRAFILES separate_compil
 RUN(NAME separate_compilation_11 LABELS gfortran llvm EXTRAFILES separate_compilation_11a.f90)
 RUN(NAME separate_compilation_12 LABELS gfortran llvm EXTRAFILES separate_compilation_12a.f90)
 RUN(NAME separate_compilation_13 LABELS gfortran llvm EXTRAFILES separate_compilation_13a.f90)
-
+RUN(NAME separate_compilation_14 LABELS gfortran llvm EXTRAFILES separate_compilation_14a.f90)
+RUN(NAME separate_compilation_15 LABELS gfortran llvm EXTRAFILES separate_compilation_15a.f90)
 
 
 

--- a/integration_tests/separate_compilation_14.f90
+++ b/integration_tests/separate_compilation_14.f90
@@ -1,0 +1,6 @@
+program separate_compilation_14a
+use separate_compilation_14a_module
+print *, global_var
+print *, global_var_initialised
+if (global_var_initialised /= 89) error stop
+end program separate_compilation_14a

--- a/integration_tests/separate_compilation_14a.f90
+++ b/integration_tests/separate_compilation_14a.f90
@@ -1,0 +1,4 @@
+module separate_compilation_14a_module
+integer :: global_var
+integer :: global_var_initialised = 89
+end module separate_compilation_14a_module

--- a/integration_tests/separate_compilation_15.f90
+++ b/integration_tests/separate_compilation_15.f90
@@ -1,0 +1,7 @@
+program separate_compilation_15a
+
+use, non_intrinsic :: separate_compilation_15a_module
+
+call test_solver()
+
+end program test

--- a/integration_tests/separate_compilation_15.f90
+++ b/integration_tests/separate_compilation_15.f90
@@ -4,4 +4,4 @@ use, non_intrinsic :: separate_compilation_15a_module
 
 call test_solver()
 
-end program test
+end program separate_compilation_15a

--- a/integration_tests/separate_compilation_15a.f90
+++ b/integration_tests/separate_compilation_15a.f90
@@ -1,0 +1,39 @@
+module separate_compilation_15a_module
+implicit none
+private
+public :: test_solver
+contains
+
+subroutine test_solver()
+
+real :: xl(2)
+real :: xu(2)
+
+call recursive_fun2()
+
+print *, "xl = ", xl
+if ( abs(sum(xl) - (-7.0)) > 1e-8 ) error stop
+print *, "xu = ", xu
+if ( abs(sum(xu) - 2.0) > 1e-8 ) error stop
+
+contains
+
+subroutine take_xl_xu(xl, xu)
+
+implicit none
+real, intent(out) :: xl(:)
+real, intent(out) :: xu(:)
+
+xl = [12.0, -19.0]
+xu = [1.0, 1.0]
+end subroutine
+
+subroutine recursive_fun2()
+implicit none
+call take_xl_xu(xl, xu)
+end subroutine recursive_fun2
+
+end subroutine test_solver
+
+
+end module separate_compilation_15a_module

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -3049,6 +3049,53 @@ public:
             [](llvm::Constant* elem) { return elem->isNullValue(); });
     }
 
+    void set_global_variable_linkage_as_common(llvm::Value* ptr, ASR::abiType x_abi) {
+        if ( compiler_options.generate_object_code ) {
+            /*
+                In case of global variables without initialization, clang
+                generates a common symbol. For the following C code:
+
+                ```
+                int global_var;
+                int global_variable_initalised = 42;
+                ```
+
+                on using `clang -S -emit-llvm` we get:
+
+                ```
+                ; ModuleID = 'a.c'
+                source_filename = "a.c"
+                target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
+                target triple = "arm64-apple-macosx14.0.0"
+
+                @global_var_initialized = global i32 42, align 4
+                @global_var = common global i32 0, align 4
+
+                !llvm.module.flags = !{!0, !1, !2, !3, !4}
+                !llvm.ident = !{!5}
+
+                !0 = !{i32 2, !"SDK Version", [2 x i32] [i32 14, i32 4]}
+                !1 = !{i32 1, !"wchar_size", i32 4}
+                !2 = !{i32 8, !"PIC Level", i32 2}
+                !3 = !{i32 7, !"uwtable", i32 1}
+                !4 = !{i32 7, !"frame-pointer", i32 1}
+                !5 = !{!"Apple clang version 15.0.0 (clang-1500.3.9.4)"}
+                ```
+
+                Hence, we set the linkage to CommonLinkage for
+                global variables without initialization in case
+                `compiler_options.generate_object_code` is set to true.
+            */
+            llvm::GlobalVariable *gv = llvm::cast<llvm::GlobalVariable>(
+                ptr
+            );
+            if ( x_abi != ASR::abiType::ExternalUndefined &&
+                x_abi != ASR::abiType::BindC ) {
+                gv->setLinkage(llvm::GlobalValue::CommonLinkage);
+            }
+        }
+    }
+
     void visit_Variable(const ASR::Variable_t &x) {
         if (x.m_value && x.m_storage == ASR::storage_typeType::Parameter) {
             this->visit_expr_wrapper(x.m_value, true);
@@ -3095,6 +3142,7 @@ public:
                     module->getNamedGlobal(llvm_var_name)->setInitializer(
                             llvm::ConstantInt::get(context,
                                 llvm::APInt(init_value_bits, 0)));
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3118,6 +3166,7 @@ public:
                                 llvm::ConstantFP::get(context,
                                     llvm::APFloat((double)0)));
                     }
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3137,6 +3186,7 @@ public:
                     module->getNamedGlobal(llvm_var_name)->setInitializer(initializer);
                 } else {
                     module->getNamedGlobal(llvm_var_name)->setInitializer(llvm::ConstantArray::getNullValue(type));
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3151,6 +3201,7 @@ public:
                     module->getNamedGlobal(llvm_var_name)->setInitializer(
                             llvm::ConstantInt::get(context,
                                 llvm::APInt(1, 0)));
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3165,6 +3216,7 @@ public:
                     module->getNamedGlobal(llvm_var_name)->setInitializer(
                             llvm::Constant::getNullValue(character_type)
                         );
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3181,6 +3233,7 @@ public:
                             llvm::ConstantPointerNull::get(
                                 static_cast<llvm::PointerType*>(void_ptr))
                             );
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3200,6 +3253,7 @@ public:
                                     llvm::ConstantPointerNull::get(
                                         static_cast<llvm::PointerType*>(void_ptr))
                                     );
+                            set_global_variable_linkage_as_common(ptr, x.m_abi);
                         }
                     }
                     llvm_symtab[h] = ptr;
@@ -3215,6 +3269,7 @@ public:
                             module->getNamedGlobal(llvm_var_name)->setInitializer(
                                     llvm::Constant::getNullValue(type)
                                 );
+                            set_global_variable_linkage_as_common(ptr, x.m_abi);
                         }
                     }
                     llvm_symtab[h] = ptr;
@@ -3230,6 +3285,7 @@ public:
                     } else {
                         module->getNamedGlobal(llvm_var_name)
                             ->setInitializer(llvm::Constant::getNullValue(type));
+                        set_global_variable_linkage_as_common(ptr, x.m_abi);
                     }
                 }
                 llvm_symtab[h] = ptr;
@@ -3278,6 +3334,7 @@ public:
                             llvm::ConstantPointerNull::get(
                                 static_cast<llvm::PointerType*>(x_ptr))
                             );
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3352,6 +3409,7 @@ public:
                     std::vector<llvm::Constant*> elements = { re, im };
                     llvm::Constant* complex_init = llvm::ConstantStruct::get(static_cast<llvm::StructType*>(type), elements);
                     module->getNamedGlobal(llvm_var_name)->setInitializer(complex_init);
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;
@@ -3368,6 +3426,7 @@ public:
                     module->getNamedGlobal(llvm_var_name)->setInitializer(
                             llvm::Constant::getNullValue(type)
                         );
+                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -3409,7 +3409,6 @@ public:
                     std::vector<llvm::Constant*> elements = { re, im };
                     llvm::Constant* complex_init = llvm::ConstantStruct::get(static_cast<llvm::StructType*>(type), elements);
                     module->getNamedGlobal(llvm_var_name)->setInitializer(complex_init);
-                    set_global_variable_linkage_as_common(ptr, x.m_abi);
                 }
             }
             llvm_symtab[h] = ptr;

